pkg/install/kubeconfig.go

```diff
@@ -0,0 +1,78 @@
+package install
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"reflect"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/tls"
+	clientcmd "k8s.io/client-go/tools/clientcmd/api/v1"
+)
+
+// generateAROServiceKubeconfig generates additional admin credentials and kubeconfig
+// based on admin kubeconfig found in graph
+func (i *Installer) generateAROServiceKubeconfig(ctx context.Context, g graph, aroServiceName string) (*kubeconfig.AdminInternalClient, error) {
+	ca := g[reflect.TypeOf(&tls.AdminKubeConfigSignerCertKey{})].(*tls.AdminKubeConfigSignerCertKey)
+	cfg := &tls.CertCfg{
+		Subject:      pkix.Name{CommonName: aroServiceName, Organization: []string{"system:masters"}},
+		KeyUsages:    x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		Validity:     tls.ValidityTenYears,
+	}
+
+	var clientCertKey tls.AdminKubeConfigClientCertKey
+
+	// generated key and cert files are stored as filenameBase.key and filenameBase.crt
+	filenameBase := strings.ReplaceAll(":", "-", aroServiceName)
+
+	err := clientCertKey.SignedCertKey.Generate(cfg, ca, filenameBase, tls.DoNotAppendParent)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a Config for the new service kubeconfig based on the generated cluster admin Config
+	adminInternalClient := g[reflect.TypeOf(&kubeconfig.AdminInternalClient{})].(*kubeconfig.AdminInternalClient)
+	aroServiceInternalClient := kubeconfig.AdminInternalClient{}
+	aroServiceInternalClient.Config = &clientcmd.Config{
+		Clusters: adminInternalClient.Config.Clusters,
+		AuthInfos: []clientcmd.NamedAuthInfo{
+			{
+				Name: aroServiceName,
+				AuthInfo: clientcmd.AuthInfo{
+					ClientCertificateData: clientCertKey.CertRaw,
+					ClientKeyData:         clientCertKey.KeyRaw,
+				},
+			},
+		},
+		Contexts: []clientcmd.NamedContext{
+			{
+				Name: aroServiceName,
+				Context: clientcmd.Context{
+					Cluster:  adminInternalClient.Config.Contexts[0].Context.Cluster,
+					AuthInfo: aroServiceName,
+				},
+			},
+		},
+		CurrentContext: aroServiceName,
+	}
+
+	data, err := yaml.Marshal(aroServiceInternalClient.Config)
```

Comment:
Didn't we agree that generateAROServiceKubeconfig would return `[]byte, error`?

Reply:
we have discussed this, but I am not convinced that this is necessarily better. In the way I suggest the function returns a complete structure, which aids with readability and usability. Especially that to make it a complete structure all it takes is just another text field. 
It also makes it clearer to people who read this code where this function is called:
```
		i.doc, err = i.db.PatchWithLease(ctx, i.doc.Key, func(doc *api.OpenShiftClusterDocument) error {
			// used for the SAS token with which the bootstrap node retrieves its
			// ignition payload
			doc.OpenShiftCluster.Properties.Install.Now = time.Now().UTC()
			doc.OpenShiftCluster.Properties.AdminKubeconfig = adminInternalClient.File.Data
			doc.OpenShiftCluster.Properties.AROServiceKubeconfig = aroServiceInternalClient.File.Data
			return nil
		})
```
in this way people can see that this is the same structure as the other one and they don't need to jump into the function to try to figure out what is that value that is returned.